### PR TITLE
Non-conjugate inference with autodiff and importance sampling

### PR DESCRIPTION
**Pull request by semihakbayrak**
Nonconjugate inference for univariate Gaussian priors is allowed by implementing a node called "nonconjugate" that calculate logpdf messages by change of variables and autodiff, and abstract distribution type that computes some sufficient statistics by ancestral sampling. Logpdf messages are employed in posterior inference approximation and abstract distribution allows VMP and free energy calculations to be carried out. Posterior inference is approximated by importance sampling to avoid hyper-parameters, such as learning rate, required in optimisation based inference methods.

### CODE REVIEW

File: src/probability_distribution.jl [hunk 1]
```diff
@@ -4,6 +4,7 @@ ProbabilityDistribution,
 Univariate,
 Multivariate,
 MatrixVariate,
+Munivariate,
```
Comment (ivan-bocharov) [inline src/probability_distribution.jl]: This type was added in order to tell FL that a node's function maps a `Multivariate` distribution to a `Univariate` distribution. FL already supports update rules with `Multivariate` input(s) and `Univariate` output.

**Comment by ThijsvdLaar** [comment]
Thanks for this great contribution.

Still, I have a problem with the way this PR handles the replacement of the the `Nonlinear` node. I understand the need for the separation of the model and inference specification. However, in the end it comes down to two (sub-optimal) approaches: (1) explicitly include the new node type in the model definition and (2) swapping the node later during the inference specification. Both alter the model definition in order to handle an inference problem. The difference is that (1) is explicit about it, while (2) sweeps it under the rug and hopes nobody notices. I'd rather be explicit and upfront about our limitations in handling this inference issue.

**Comment by ivan-bocharov** [comment]
Thank you for your feedback. I have tried to address the concerns you have outlined in your comment.
